### PR TITLE
[mongodb] Upgrade DB driver, add more type handlings, fix QuantityType handling

### PR DESCRIPTION
**Pull request by ulbi**
# [mongodb] Upgrade DB driver, add more type handlings, fix QuanitityType handling

# Description

This pull request addresses issues #16308 and #16310, introducing significant updates and improvements to the system:

## PR Classification
- **Improvement**: Updates the MongoDB driver for enhanced compatibility and adds comprehensive unit testing.
- **Bugfix**: Fixes handling for various item types, ensuring consistent behavior across different MongoDB versions.

## Goals and Motivation
The main objective is to modernize the MongoDB driver used within our system from the outdated `mongo-java-driver 2.13.1` (2015) to the current `mongo-driver-sync 4.11.1`. This upgrade not only extends support to MongoDB versions 3.6 through 6.0 but also addresses compatibility issues with MongoDB 5.0 and above, which is no longer adequately supported by the old driver.

## Noteworthy Changes for End Users
- **Database Driver Upgrade**: Transition to `mongo-driver-sync 4.11.1`, enhancing support for MongoDB versions 3.6 to 6.0 and ensuring better system performance and reliability.
- **Enhanced Item Support**: Improved handling for `NumberItem` with `QuantityTypes` and added support for `ColorItem`, `LocationItem`, `ImageItem`, `PlayerItem`, and `CallItem`, along with their respective state types, enhancing the system's versatility and user experience.

## Unit Testing Enhancements
- Added unit tests to validate type storage and querying, ensuring robustness and reliability.
- Expanded testing to cover all supported MongoDB versions (3.6, 4.4, 5.0, 6.0), guaranteeing compatibility across different database versions.

## Technical Improvements
- For consistency with InfluxDB and compatibility with charting tools like Grafana, `QuantityType` handling now stores only the number without the unit. This adjustment allows for more seamless integration with external tools that may not support string-based data types.

## Documentation and Guidelines
- Updated documentation to reflect these changes, ensuring users and developers can easily understand and adopt the new features and improvements.

### CODE REVIEW

File: bundles/org.openhab.persistence.mongodb/pom.xml [hunk 1]
```diff
@@ -14,12 +14,84 @@
...
+      <artifactId>zstd-jni</artifactId>
+      <version>1.5.5-3</version>
+      <scope>compile</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.lastnpe.eea</groupId>
+      <artifactId>eea-all</artifactId>
+      <version>2.2.1</version>
+      <scope>compile</scope>
+    </dependency>
```
Comment (J-N-K) [inline bundles/org.openhab.persistence.mongodb/pom.xml:57]: I don't think you need this one.

File: bundles/org.openhab.persistence.mongodb/pom.xml [hunk 2]
```diff
@@ -14,12 +14,84 @@
...
+      <artifactId>logback-classic</artifactId>
+      <version>1.2.9</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.mockito</groupId>
+      <artifactId>mockito-core</artifactId>
+      <version>4.2.0</version>
+      <scope>test</scope>
+    </dependency>
```
Comment (J-N-K) [inline bundles/org.openhab.persistence.mongodb/pom.xml:83]: These should be available from core

**Review by J-N-K — COMMENTED** [review]
Regarding the change: I fear that all stored quantity-type will be  lost, is that correct? Do we have a chance to convert them or a "compatibilty mode" for accessing them?

**Comment by ulbi** [comment]
@J-N-K @jlaur just created this PR to address a couple of challenges with the current mongodb driver. I created this as draft in the first place, since I plan two more things:

1. add code to also handle the data stored by the current mongodb persistence as fas as possible e.g. QuantityType (currently stored as string, but not restored properly) and all other where "toString" provides the full context (ColorItem, LocationItem, ...) - this will not work for ImageItem, since a toString of RawType does not store everything
2. add exception handling to catch wrong data from the database
3. add exception handling if the ImageItem or anything else tries to store more than 16 MB (document size limit from MongoDB)
4. store the Unit for QuantityType if provided in a separate field and then restore it from there.
5. Address your comments ;-)

Once done, this should cover a couple of issues.

**Comment by ulbi** [comment]
> Regarding the change: I fear that all stored quantity-type will be lost, is that correct? Do we have a chance to convert them or a "compatibilty mode" for accessing them?

You are too fast! I will add something for this as well, just wanted to create the PR already, since there are quite some changes already. And just figured out, that the online build environment does not contain docker, so all tests using "Testcontainers" do fail - too bad, this is a great way to verify the different mongodb versions.

**Comment by ulbi** [comment]
@J-N-K Just a quick question, anything else required to get this merged? I'm asking because the changes required for #16338 are ready as well, should I put it in the PR as well?

**Comment by lsiepel** [comment]
> @J-N-K Just a quick question, anything else required to get this merged? I'm asking because the changes required for #16338 are ready as well, should I put it in the PR as well?

As I don’t expect many users to test this persistence prior to stable. Maybe good to integrate it and get some mileage on both changes.

File: bundles/org.openhab.persistence.mongodb/src/main/java/org/openhab/persistence/mongodb/internal/MongoDBTypeConversions.java [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+                return "$gte";
+            case LT:
+                return "$lt";
+            case LTE:
+                return "$lte";
+            case NEQ:
+                return "$neq";
+            default:
+                return null;
+        }
```
Comment (J-N-K) [inline bundles/org.openhab.persistence.mongodb/src/main/java/org/openhab/persistence/mongodb/internal/MongoDBTypeConversions.java:206]:
```suggestion
        return switch (operator) {
            case EQ ->"$eq";
            case GT -> "$gt";
            case GTE -> "$gte";
            case LT -> "$lt";
            case LTE -> "$lte";
            case NEQ -> "$neq";
            default -> null;
        };
```

File: bundles/org.openhab.persistence.mongodb/src/test/java/org/openhab/persistence/mongodb/internal/MongoDBPersistenceServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,802 @@
...
+import ch.qos.logback.classic.spi.ILoggingEvent;
+import ch.qos.logback.core.read.ListAppender;
+import de.bwaldvogel.mongo.backend.memory.MemoryBackend;
+
+/**
+ * This is the implementation of the test for MongoDB {@link PersistenceService}.
+ *
+ * @author René Ulbricht - Initial contribution
+ */
+public class MongoDBPersistenceServiceTest {
```
Comment (J-N-K) [inline bundles/org.openhab.persistence.mongodb/src/test/java/org/openhab/persistence/mongodb/internal/MongoDBPersistenceServiceTest.java:54]:
```suggestion
@NonNullByDefault
public class MongoDBPersistenceServiceTest {
```

File: bundles/org.openhab.persistence.mongodb/src/test/java/org/openhab/persistence/mongodb/internal/DatabaseTestContainer.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+import de.bwaldvogel.mongo.backend.memory.MemoryBackend;
+
+/**
+ * This class provides a container for MongoDB for testing purposes.
+ * It uses the Testcontainers library to manage the MongoDB container.
+ * It also provides an in-memory MongoDB server for testing.
+ * 
+ * @author René Ulbricht - Initial contribution
+ */
+public class DatabaseTestContainer {
```
Comment (J-N-K) [inline bundles/org.openhab.persistence.mongodb/src/test/java/org/openhab/persistence/mongodb/internal/DatabaseTestContainer.java:31]:
```suggestion
@NonNullByDefault
public class DatabaseTestContainer {
```

File: bundles/org.openhab.persistence.mongodb/src/test/java/org/openhab/persistence/mongodb/internal/DataCreationHelper.java [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+import ch.qos.logback.classic.spi.ILoggingEvent;
+import ch.qos.logback.core.read.ListAppender;
+import de.bwaldvogel.mongo.backend.memory.MemoryBackend;
+
+/**
+ * This class provides helper methods to create test items.
+ * 
+ * @author René Ulbricht - Initial contribution
+ */
+public class DataCreationHelper {
```
Comment (J-N-K) [inline bundles/org.openhab.persistence.mongodb/src/test/java/org/openhab/persistence/mongodb/internal/DataCreationHelper.java:64]:
```suggestion
@NonNullByDefault
public class DataCreationHelper {
```

File: bundles/org.openhab.persistence.mongodb/src/main/java/org/openhab/persistence/mongodb/internal/MongoDBTypeConversions.java [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class handles the conversion of types between openHAB and MongoDB.
+ * It provides methods to convert openHAB states to MongoDB compatible types and vice versa.
+ * It also provides a method to convert openHAB filter operators to MongoDB query operators.
+ *
+ * @author René Ulbricht - Initial contribution
+ */
+public class MongoDBTypeConversions {
```
Comment (J-N-K) [inline bundles/org.openhab.persistence.mongodb/src/main/java/org/openhab/persistence/mongodb/internal/MongoDBTypeConversions.java:46]:
```suggestion
@NonNullByDefault
public class MongoDBTypeConversions {
```

File: bundles/org.openhab.persistence.mongodb/src/test/java/org/openhab/persistence/mongodb/internal/SetupResult.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+import org.osgi.framework.BundleContext;
+
+import com.mongodb.client.MongoDatabase;
+
+/**
+ * This class provides helper methods to create test items.
+ * 
+ * @author René Ulbricht - Initial contribution
+ */
+public class SetupResult {
```
Comment (J-N-K) [inline bundles/org.openhab.persistence.mongodb/src/test/java/org/openhab/persistence/mongodb/internal/SetupResult.java:27]:
```suggestion
@NonNullByDefault
public class SetupResult {
```

File: bundles/org.openhab.persistence.mongodb/src/test/java/org/openhab/persistence/mongodb/internal/VerificationHelper.java [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+
+/**
+ * This is a helper class for verifying various aspects of the MongoDB persistence service.
+ * It provides methods for verifying log messages, MongoDB documents, and query results.
+ * Each verification method checks if the actual value matches the expected value and throws an
+ * AssertionError if they do not match.
+ *
+ * @author René Ulbricht - Initial contribution
+ */
+public class VerificationHelper {
```
Comment (J-N-K) [inline bundles/org.openhab.persistence.mongodb/src/test/java/org/openhab/persistence/mongodb/internal/VerificationHelper.java:43]:
```suggestion
@NonNullByDefault
public class VerificationHelper {
```

**Review by J-N-K — CHANGES_REQUESTED** [review]

**Comment by ulbi** [comment]
@lsiepel Can do it, this will fix a couple of the requested changes as well, since I fixed most of the compiler warnings in there. Will just put it in here with the changes requested by @J-N-K .

File: bundles/org.openhab.persistence.mongodb/src/main/java/org/openhab/persistence/mongodb/internal/MongoDBTypeConversions.java [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+            }
+
+            if (value instanceof String) {
+                return new QuantityType<>(value.toString());
+            }
+            if (unit != null) {
+                return new QuantityType<>(((Number) value).doubleValue(), unit);
+            } else {
+                return new DecimalType(((Number) value).doubleValue());
+            }
```
[the remark this answers is not in the document]
Comment (ulbi) [reply bundles/org.openhab.persistence.mongodb/src/main/java/org/openhab/persistence/mongodb/internal/MongoDBTypeConversions.java:108]: Done it. In order to increase readability I also move the inlinemethods to single private methods. This seems to be better

**Comment by ulbi** [comment]
@J-N-K @lsiepel I should have covered all comments and added the enhancement for #16338 as well.

**Review by J-N-K — APPROVED** [review]
Thanks!

**Pull request merged**